Adapt to undertaker v1.0.0

sttk opened the pull request:
This PR is for adapting to both v1.0.0 and previous versions of `undertaker`.
(And this adds a test case for 100% coverage.)

phated commented on appveyor.yml:
```diff
@@ -15,8 +15,8 @@ environment:
 install:
-  - IF %nodejs_version% EQU 0.10 npm -g install npm@2
-  - IF %nodejs_version% EQU 0.10 set PATH=%APPDATA%\npm;%PATH%
+  - npm -g install npm@latest
```
are you sure you want to be using npm 3?  It's so much slower, and appveyor is already extremely slow.

sttk replied:
npm install EPERM errors frequently occured on appveyor, and it seems that this error can be fixed by using npm3.

ref. https://github.com/npm/npm/issues/9696

phated commented on lib/shared/log/tasks.js:
```diff
@@ -13,7 +13,23 @@ function isObject(val) {
   return typeof val === 'object' && !Array.isArray(val);
 }
 
+function wrapGetTask(getTask) {
```
this probably shouldn't be in the shared module and should instead go in the `getTask` methods of alpha.2 and 4.0

sttk replied:
I got it. I'll modify so.

phated commented on lib/shared/log/tasks.js:
```diff
@@ -13,7 +13,23 @@ function isObject(val) {
+}
+
+function createTask(task) {
+  var unwrapped = task.unwrap ? task.unwrap() : {};
```
1. try to read `task.description` before unwrapping
2. if it is empty, check for `typeof task.unwrap === 'function'`, unwrap and check for `unwrapped.description`

sttk replied:
I got it. I'll modify so.

sttk replied:
Which do you think best to check if `task.description` is not empty (or specified)?

```js
 1. if (task.description) ...
 2. if (task.description != null) ...
 3. if (typeof task.description === 'string') ...
 4. if ('description' in task) ...
```

phated replied:
I think `typeof task.description === 'string'` is best because we don't handle any other cases.

sttk replied:
Thanks. I got it. I'll do the same by using `lodash`

phated commented on package.json:
```diff
@@ -27,7 +27,7 @@
     "prepublish": "marked-man --name gulp docs/CLI.md > gulp.1",
     "pretest": "npm run lint",
-    "test": "lab test/*.js -cv",
+    "test": "lab test/*.js -cv -I Reflect",
```
yeah, this is annoying.  These tests will need to be switch to mocha like all the other repositories.  This is a fine fix for now.

sttk commented:
Sorry. This push is a mistake. Please wait a little more.

sttk commented:
I've finished modifying the codes.
I used `lodash` to check the properties and modified the same check in `tasks.js`, too.
In addition, I've added node 6 to the test target.

phated commented on lib/shared/log/tasks.js:
```diff
@@ -3,15 +3,11 @@
-function isObject(val) {
-  return typeof val === 'object' && !Array.isArray(val);
-}
+var _ = require('lodash');
```
I think people were trying to transition all the gulp dependencies to use `lodash.*` packages, so this should require `lodash.sortby`, `lodash.isstring` and `lodash.isplainobject`

phated commented on lib/versioned/^4.0.0/log/getTask.js:
```diff
@@ -0,0 +1,39 @@
+function getTask(gulpInst) {
+  return function(name) {
+    var task = gulpInst.task(name);
+    var func = isFunction(task.unwrap) ? task.unwrap() : {};
```
this can be moved to the `getDescription` and `getFlags` functions.  It will end up calling `unwrap` twice in some scenarios but that shouldn't be much overhead and it avoids it if the properties are on the wrapper

phated commented on package.json:
```diff
@@ -37,7 +37,7 @@
     "interpret": "^1.0.0",
     "liftoff": "^2.1.0",
-    "lodash.sortby": "^4.0.1",
+    "lodash": "^4.13.1",
```
avoid using the full lodash package, instead use the small modules

phated commented on test/fixtures/flags-tasks-with-nowrap.js:
```diff
@@ -0,0 +1,40 @@
+gulp.task('task1').description = 'Description for gulp.task("task1")';
+
+gulp.task('task2', function() {});
+if (!gulp.task('task2').unwrap) {
```
instead of all this, can't you just use a named function?

sttk replied:
The purpose of this test is that `gulp-cli` supports both the old way, `description` is gotten by `gulp.task(name)`, and the new way, it is gotten by `gulp.task(name).unwrap()`.

This cannot be controlled by setting using named function.

And getting task tree correctly by named functions is tested by `flags-tasks-with-desc.js`.

phated commented on test/fixtures/flags-tasks-with-nowrap.js:
```diff
@@ -0,0 +1,40 @@
+
+gulp.task('task3', function() {});
+gulp.task('task3', function() {});
+if (!gulp.task('task3').unwrap) {
```
instead of all this, can't you just use a named function?

phated commented on test/fixtures/flags-tasks-with-nowrap.js:
```diff
@@ -0,0 +1,40 @@
+gulp.task('task2').unwrap()
+  .description = 'Description for gulp.task("task2").unwrap()';
+
+gulp.task('task3', function() {});
```
why is `task3` set twice?

sttk replied:
It might be my mistake or I couldn't remember the reaseon. I'll remove one line of them and one more line about 'no-desc'.

phated commented on test/fixtures/flags-tasks-with-nowrap.js:
```diff
@@ -0,0 +1,40 @@
+
+gulp.task('no-desc', function() {});
+gulp.task('no-desc', function() {});
+if (!gulp.task('no-desc').unwrap) {
```
instead of all this, can't you just use a named function?

phated commented on test/taskTree.js:
```diff
@@ -48,4 +48,58 @@ lab.experiment('taskTree()', function() {
     done();
   });
+
+  lab.test('processes children recursively.', function(done) {
```
where did this test come from?

sttk replied:
This test is what I forgot to include in previous PR and for coverage 100%.

phated commented:
@sttk sorry it took me so long to review.  I've commented on some things I noticed.

sttk commented:
@phated Don't worry. I know you are busy with other projects. Thank you for reviewing.

sttk commented:
@phated I've modified what you pointed out. And I've merged the tests for the flag `--tasks` in one.

phated commented:
Published as 1.2.2

phated commented:
Excellent work @sttk!  Really glad you caught this oversight in my undertaker changes.
